### PR TITLE
Adding RHACS(StackRox) UDI connector

### PR DESCRIPTION
**Pull request by lakshmi51974368**
Adding Red Hat Advanced Cluster Security RHACS(StackRox) UDI connector.

### CODE REVIEW

File: stix_shifter_modules/rhacs/configuration/config.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+    "connection": {
+        "type": {
+            "displayName": "RHACS(StackRox)"
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/configuration/config.json:4]: better to set the full name instead.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+        return value
+
+    def _eval_combined_comparison_exp(self, expression):
+        """
+        Function for parsing combined comparison expression
+        :param expression: expression object
+        """
+        operator = self._lookup_comparison_operator(expression.operator)
+        if operator == '+':
+            operator = '%2B'
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_translation/query_constructor.py:388]: you should encode characters in python way not hardcode it.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+    translate stix pattern to native data source query language
+    """
+
+    def __init__(self, pattern: Pattern, data_model_mapper, options):
+        logger.info("RHACS Connector")
+        self.dmm = data_model_mapper
+        self.options = options
+        self.timeframe = []
+        self.comparator_lookup = self.dmm.map_comparator()
+        self.type_map = self.load_json(TYPE_MAP_PATH)
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_translation/query_constructor.py:40]: use absolute path to read files from a different folder.

File: stix_shifter_modules/rhacs/stix_transmission/connector.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+                            len(inner_dict.get('violations')) > 0:
+                        for container in dt['Container']:
+                            # exit if result limit crossed
+                            if total_count > max_range:
+                                break
+                            for item in inner_dict.get('violations'):
+                                if item.get('networkFlowInfo') is not None:
+                                    is_process_or_netflow = True
+                                    if item['time'] is not None:
+                                        commondataset["firstObserved"] = item['time']
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_transmission/connector.py:79]: why do you need to modify the time datasource fields to firstObserved at this point?

**Review by mdazam1942 — COMMENTED** [review]
I see a lot of results schema modification in stix_shifter_modules/rhacs/stix_transmission/connector.py. Are all of those necessary? Is  to_stix mapping not enough for the results translation?  Totally changing the results schema is not a good idea if we need to look at raw data for missing fields in future.

File: stix_shifter_modules/rhacs/configuration/config.json (hunk 1 above)
Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/configuration/config.json]: Updated as per suggestion.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py (hunk 2 above)
Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_translation/query_constructor.py]: Updated as per suggestion.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py (hunk 3 above)
Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_translation/query_constructor.py:41]: Updated load_json function to use absolute path as suggested.

File: stix_shifter_modules/rhacs/stix_transmission/connector.py (hunk 4 above)
Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_transmission/connector.py:79]: This is required to handle multiple process/netflow violation for a single alert_id. The details has been shared by email yesterday.

**Comment by lakshmi51974368** [comment]
> 

Yes, the changes in result schema is required. The details has been shared by email yesterday.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+        """
+        match = re.search(r'Violation\s*Time:', query)
+        if match:
+            final_query = query
+        else:
+            converted_timestamp = \
+                QueryStringPatternTranslator._parse_time_range(qualifier, self.options["time_range"])
+            QueryStringPatternTranslator._check_time_range_values(converted_timestamp)
+            self.timeframe += converted_timestamp
+            final_query = f'{query}%2BViolation Time:>=' \
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_translation/query_constructor.py:303]: encoded character `%2B` is still hardcoded here.

File: stix_shifter_modules/rhacs/README.md [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+
+#### STIX Translate query
+```shell
+translate rhacs query '{}' "[x-rhacs-cluster:name = 'xxxx'] START t'2022-07-07T11:00:00.000Z' STOP t'2022-07-08T11:00:00.003Z'"
+```
+#### STIX Translate query - output
+```json
+{
+    "queries": [
+        "Cluster:\"xxxx\"%2BViolation Time:>=07/07/2022"
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/README.md:36]: is there any reason why query translation totally disregard the hr/mins/seconds from the time ranges?

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py (hunk 5 above)
Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_translation/query_constructor.py]: Updated to encode '+' character.

File: stix_shifter_modules/rhacs/stix_translation/query_constructor.py [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+        return value
+
+    def _eval_combined_comparison_exp(self, expression):
+        """
+        Function for parsing combined comparison expression
+        :param expression: expression object
+        """
+        operator = self._lookup_comparison_operator(expression.operator)
+        if operator == '+':
+            operator = quote(operator)
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_translation/query_constructor.py:388]: the url encoding should be done at the transmit call not at query constructions. please move the encoding function to stix_transmission unless you have any specific reason to keep it here.

Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_translation/query_constructor.py]: As suggested the url encoding is done in transmit.

File: stix_shifter_modules/rhacs/stix_translation/json/stix_2_1/to_stix_map.json [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    "execFilePath": {
+      "key": "process.cwd",
+      "object": "proc"
+    },
+    "time": {
+      "key": "process.created",
+      "object": "proc"
+    },
+    "id": {
+      "key": "process.extensions.x-rhacs-process-ext.x_unique_id",
```
Comment (mdazam1942) [inline stix_shifter_modules/rhacs/stix_translation/json/stix_2_1/to_stix_map.json:156]: as discussed, this needs be taken out of the extensions. 
`"key": "process.x_unique_id"`

Comment (lakshmi51974368) [reply stix_shifter_modules/rhacs/stix_translation/json/stix_2_1/to_stix_map.json]: updated as suggested.

**Review by mdazam1942 — APPROVED** [review]

**Pull request merged**